controllers/machinepool_controller_phases.go

```diff
@@ -0,0 +1,232 @@
+	if !loaded && r.controller != nil {
+		klog.Infof("Adding watcher on external object %q", obj.GroupVersionKind())
+		err := r.controller.Watch(
+			&source.Kind{Type: obj},
+			&handler.EnqueueRequestForOwner{OwnerType: &clusterv1.MachinePool{}},
+		)
+		if err != nil {
+			r.externalWatchers.Delete(obj.GroupVersionKind().String())
+			return nil, errors.Wrapf(err, "failed to add watcher on external object %q", obj.GroupVersionKind())
+		}
+	}
+
+	// Set error reason and message, if any.
+	errorReason, errorMessage, err := external.FailuresFrom(obj)
+	if err != nil {
+		return nil, err
+	}
+	if errorReason != "" {
+		machinePoolStatusError := capierrors.MachinePoolStatusError(errorReason)
+		mp.Status.FailureReason = &machinePoolStatusError
+	}
+	if errorMessage != "" {
+		mp.Status.FailureMessage = pointer.StringPtr(
+			fmt.Sprintf("Failure detected from referenced resource %v with name %q: %s",
+				obj.GroupVersionKind(), obj.GetName(), errorMessage),
+		)
+	}
+
+	return obj, nil
+}
+
+// reconcileBootstrap reconciles the Spec.Bootstrap.ConfigRef object on a MachinePool.
+func (r *MachinePoolReconciler) reconcileBootstrap(ctx context.Context, mp *clusterv1.MachinePool) error {
+	// TODO(vincepri): Move this validation in kubebuilder / webhook.
+	if mp.Spec.Template.Spec.Bootstrap.ConfigRef == nil && mp.Spec.Template.Spec.Bootstrap.Data == nil {
+		return errors.Errorf(
+			"Expected at least one of `Bootstrap.ConfigRef` or `Bootstrap.Data` to be populated for MachinePool %q in namespace %q",
+			mp.Name, mp.Namespace,
+		)
+	}
+
+	// Call generic external reconciler if we have an external reference.
+	var bootstrapConfig *unstructured.Unstructured
+	if mp.Spec.Template.Spec.Bootstrap.ConfigRef != nil {
+		var err error
+		bootstrapConfig, err = r.reconcileExternal(ctx, mp, mp.Spec.Template.Spec.Bootstrap.ConfigRef)
+		if err != nil {
+			return err
+		}
+	}
+
+	// If the bootstrap config is being deleted, return early.
+	if !bootstrapConfig.GetDeletionTimestamp().IsZero() {
+		return nil
+	}
+
+	// Determine if the bootstrap provider is ready.
+	ready, err := external.IsReady(bootstrapConfig)
+	if err != nil {
+		return err
+	}
+
+	mp.Status.BootstrapReady = ready
+	if !mp.Status.BootstrapReady {
+		return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: externalReadyWait},
+			"Bootstrap provider for MachinePool %q in namespace %q is not ready, requeuing", mp.Name, mp.Namespace)
+	}
+
+	// Get and set data from the bootstrap provider.
+	data, _, err := unstructured.NestedString(bootstrapConfig.Object, "status", "bootstrapData")
+	if err != nil {
+		return errors.Wrapf(err, "failed to retrieve data from bootstrap provider for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	} else if data == "" {
+		return errors.Errorf("retrieved empty data from bootstrap provider for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	mp.Spec.Template.Spec.Bootstrap.Data = pointer.StringPtr(data)
```

Comment:
We need to get the bootstrap data from the secret here, instead of the field in plaintext, which has been deprecated in v1alpha3